crates/pera-core/src/signature_mpc/mpc_manager.rs

```diff
@@ -0,0 +1,451 @@
+use pera_types::event::Event;
+use pera_types::messages_consensus::ConsensusTransaction;
+
+use pera_types::committee::EpochId;
+use rand_core::OsRng;
+use rayon::prelude::*;
+use schemars::_private::NoSerialize;
+use serde::{Deserialize, Serialize};
+use std::cmp::PartialEq;
+use std::collections::{HashMap, HashSet, VecDeque};
+use std::future::Future;
+use std::io::Write;
+use std::marker::PhantomData;
+use std::sync::{Arc, Weak};
+use std::time::Duration;
+use std::{io, mem};
+use tokio::sync::{mpsc, Mutex, RwLock};
+use tokio::time::sleep;
+use tracing::{debug, error, info};
+
+/// The message a validator can send to the other parties while running a signature MPC session.
+#[derive(Clone)]
+struct SignatureMPCMessage {
+    /// The serialized message
+    message: Vec<u8>,
+    /// The authority that sent the message
+    authority: AuthorityName,
+}
+
+/// A wrapper for the generic Party trait that allows creating new instances of the Party from only the threshold.
+/// Should be implemented internally in newer versions of the [`proof`] crate.
+pub trait CreatableParty: Advance {
+    /// The MPC Manager will create a new mpc instance after the init event is received.
+    type InitEvent: MPCEvent + Serialize + for<'a> Deserialize<'a>;
+    /// The MPC Manager will finalize the mpc instance after the finalize event is received.
+    type FinalizeEvent: MPCEvent + Serialize + for<'a> Deserialize<'a>;
+
+    fn new(parties: HashSet<PartyID>, party_id: PartyID) -> Self;
+}
+
+/// Convert a given authority name (address) to it's corresponding party ID.
+/// The party ID is the index of the authority in the committee.
+fn authority_name_to_party_id(
+    authority_name: AuthorityName,
+    epoch_store: &AuthorityPerEpochStore,
+) -> PeraResult<PartyID> {
+    Ok(epoch_store
+        .committee()
+        .authority_index(&authority_name)
+        // This should never happen, as the validator only accepts messages from committee members
+        .ok_or_else(|| {
+            PeraError::InvalidCommittee(
+                "Received a proof MPC message from a validator that is not in the committee"
+                    .to_string(),
+            )
+        })? as PartyID)
+}
+
+/// A Signature MPC session instance
+/// It keeps track of the status of the session, the channel to send messages to the instance,
+/// and the messages that are pending to be sent to the instance.
+struct SignatureMPCInstance<P: CreatableParty> {
+    status: MPCSessionStatus<P::OutputValue>,
+    /// The messages that are pending to be executed while advancing the instance
+    /// We need to accumulate threshold of those before advancing the instance
+    pending_messages: HashMap<PartyID, P::Message>,
+    consensus_adapter: Arc<dyn SubmitToConsensus>,
+    epoch_store: Weak<AuthorityPerEpochStore>,
+    epoch_id: EpochId,
+    /// The total number of parties in the chain
+    /// We can calculate the threshold and parties IDs (indexes) from it
```

Comment:
How? elaborate...

Reply:
Turns out there is already an issue about it.
We can't implement it right now since we are waiting for @ycscaly's implementation in the MPC algorithm.
https://github.com/dwallet-labs/dwallet-network/issues/268